arwen/elrondapi/elrondei.go

```diff
@@ -1732,10 +1647,105 @@ func getOriginalTxHash(context unsafe.Pointer, dataOffset int32) {
 	_ = arwen.WithFault(err, context, runtime.ElrondAPIErrorShouldFailExecution())
 }
 
-func areInSameShard(host arwen.VMHost, leftAddress []byte, rightAddress []byte) bool {
-	blockchain := host.Blockchain()
-	leftShard := blockchain.GetShardOfAddress(leftAddress)
-	rightShard := blockchain.GetShardOfAddress(rightAddress)
+func prepareIndirectContractCallInput(
+	host arwen.VMHost,
+	sender []byte,
+	value *big.Int,
+	gasLimit int64,
+	addressOffset int32,
+	functionOffset int32,
+	functionLength int32,
+	numArguments int32,
+	argumentsLengthOffset int32,
+	dataOffset int32,
+) (*vmcommon.ContractCallInput, error) {
+	runtime := host.Runtime()
+	metering := host.Metering()
+
+	destination, err := runtime.MemLoad(addressOffset, arwen.AddressLen)
+	if err != nil {
+		return nil, err
+	}
+
+	if !host.AreInSameShard(sender, destination) {
+		return nil, arwen.ErrSyncExecutionNotInSameShard
+	}
 
-	return leftShard == rightShard
+	function, data, actualLen, err := getArgumentsFromMemory(
+		host,
+		functionOffset,
+		functionLength,
+		numArguments,
+		argumentsLengthOffset,
+		dataOffset,
+	)
+	gasToUse := metering.GasSchedule().BaseOperationCost.DataCopyPerByte * uint64(actualLen)
+	metering.UseGas(gasToUse)
+	if err != nil {
+		return nil, err
+	}
+
+	contractCallInput := &vmcommon.ContractCallInput{
+		VMInput: vmcommon.VMInput{
+			CallerAddr:  sender,
+			Arguments:   data,
+			CallValue:   value,
+			GasPrice:    0,
+			GasProvided: metering.BoundGasLimit(gasLimit),
+		},
+		RecipientAddr: destination,
+		Function:      function,
+	}
+
+	return contractCallInput, nil
+}
+
+func getArgumentsFromMemory(
+	host arwen.VMHost,
+	functionOffset int32,
+	functionLength int32,
+	numArguments int32,
+	argumentsLengthOffset int32,
+	dataOffset int32,
+) (string, [][]byte, int32, error) {
+	runtime := host.Runtime()
+
+	function, err := runtime.MemLoad(functionOffset, functionLength)
+	if err != nil {
+		return "", nil, 0, err
+	}
+
+	argumentsLengthData, err := runtime.MemLoad(argumentsLengthOffset, numArguments*4)
+	if err != nil {
+		return "", nil, 0, err
+	}
+
+	currOffset := dataOffset
+	data, err := arwen.GuardedMakeByteSlice2D(numArguments)
+	if err != nil {
+		return "", nil, 0, err
+	}
+
+	for i := int32(0); i < numArguments; i++ {
+		currArgLenData := argumentsLengthData[i*4 : i*4+4]
+		actualLen := bytesToInt32(currArgLenData)
+
+		data[i], err = runtime.MemLoad(currOffset, actualLen)
+		if err != nil {
+			return "", nil, 0, err
+		}
+
+		currOffset += actualLen
+	}
+
+	return string(function), data, currOffset - dataOffset, nil
+}
+
+func bytesToInt32(data []byte) int32 {
```

Comment:
pretty sure this function can be taken from some library
I think this one does it:
https://golang.org/pkg/encoding/binary/#example_ByteOrder_get

Reply:
Changed to use `binary.LittleEndian.Uint32()`.